airbyte-integrations/connectors/destination-langchain/destination_langchain/document_processor.py

```diff
@@ -0,0 +1,112 @@
+        if isinstance(stream, AirbyteStream):
+            return stream.name if stream.namespace is None else f"{stream.namespace}_{stream.name}"
+        else:
+            return stream.stream if stream.namespace is None else f"{stream.namespace}_{stream.stream}"
+
+    def process(self, record: AirbyteRecordMessage) -> Tuple[List[Document], Optional[str]]:
+        """
+        Generate documents from records.
+        :param records: List of AirbyteRecordMessages
+        :return: Tuple of (List of document chunks, Natural id to delete if applicable)
+        """
+        doc = self._generate_document(record)
+        if doc is None:
+            self.logger.warning(f"Record {str(record.data)[:250]}... does not contain any text fields. Skipping.")
+            return [], None
+        chunks = self._split_document(doc)
+        id_to_delete = doc.metadata[METADATA_NATURAL_ID_FIELD] if METADATA_NATURAL_ID_FIELD in doc.metadata else None
+        return chunks, id_to_delete
+
+    def _generate_document(self, record: AirbyteRecordMessage) -> Optional[Document]:
+        relevant_fields = self._extract_relevant_fields(record)
+        if len(relevant_fields) == 0:
+            return None
+        metadata = self._extract_metadata(record)
+        text = stringify_dict(relevant_fields)
+        return Document(page_content=text, metadata=metadata)
+
+    def _extract_relevant_fields(self, record: AirbyteRecordMessage) -> dict:
+        relevant_fields = {}
+        if self.text_fields:
+            for field in self.text_fields:
+                values = dpath.util.values(record.data, field, separator=".")
+                if values and len(values) > 0:
+                    relevant_fields[field] = values
+        else:
+            relevant_fields = record.data
+        return relevant_fields
+
+    def _extract_metadata(self, record: AirbyteRecordMessage) -> dict:
+        metadata = record.data
+        if self.text_fields:
+            for field in self.text_fields:
+                try:
+                    dpath.util.delete(metadata, field, separator=".")
+                except PathNotFound:
+                    pass  # if the field doesn't exist, do nothing
+        metadata = self._normalize_metadata(metadata)
+        stream_identifier = self._stream_identifier(record)
+        current_stream = self.streams[stream_identifier]
+        metadata[METADATA_STREAM_FIELD] = stream_identifier
+        # if the sync mode is deduping, use the primary key to upsert existing records instead of appending new ones
+        if current_stream.primary_key and current_stream.destination_sync_mode == DestinationSyncMode.append_dedup:
+            # TODO support nested and composite primary keys
+            metadata[METADATA_NATURAL_ID_FIELD] = record.data[current_stream.primary_key[0][0]]
+        return metadata
+
+    def _normalize_metadata(self, metadata: dict) -> dict:
```

Comment:
nit: how about `truncate`? normalization is a specific concept in Airbyte destionations